core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java

```diff
@@ -0,0 +1,440 @@
+                }
+            }
+
+            return newJoin;
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
+        }
+
+        @Override
+        public PlanNode visitSort(SortNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
+        }
+
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, Map<Symbol, SkewedContext> context)
+        {
+            /*
+                If any symbol assignments refers to a skewed symbol then add that symbol to the set of skewed symbols.
+            * */
+            PlanNode child = node.getSource().accept(this, context);
+
+            // Symbols that will be removed by this projection.
+            Set<Symbol> outputSymbols = new HashSet<>(node.getOutputSymbols());
+            Set<Symbol> symbolsToRemove = new HashSet<>(node.getSource().getOutputSymbols());
+            symbolsToRemove.removeAll(outputSymbols);
+
+            // Add all symbols that refer to a skewed symbol.
+            node.getAssignments().entrySet()
+                    .forEach(entry -> {
+                        Expression value = entry.getValue();
+                        if (entry.getValue() instanceof SymbolReference) {
+                            Symbol skewedSymbol = Symbol.from(value);
+                            if (context.containsKey(Symbol.from(value)) && !context.containsKey(entry.getKey())) {
+                                context.put(entry.getKey(), context.get(skewedSymbol));
+                            }
+                        }
+                    });
+
+            // Remove symbols that are dropped in the projection
+            context.keySet().removeAll(symbolsToRemove);
+
+            return replaceChildren(node, ImmutableList.of(child));
+        }
+
+        @Override
+        protected PlanNode visitPlan(PlanNode node, Map<Symbol, SkewedContext> context)
+        {
+            // For all other nodes, don't let the context pass through.
+            List<PlanNode> children = node.getSources().stream()
+                    .map(child -> child.accept(this, new HashMap<>()))
+                    .collect(Collectors.toList());
+            return replaceChildren(node, children);
+        }
+
+        @Override
+        public PlanNode visitTableScan(TableScanNode node, Map<Symbol, SkewedContext> context)
+        {
+            final String qualifiedTableName = metadata.getTableMetadata(session, node.getTable()).getQualifiedName().toString();
+
+            if (this.skewedTables.containsKey(qualifiedTableName)) {
```

Comment:
This doesn't handle case sensitivity issues or quoted identifiers (e.g., what happens if the table name contains a special character?)

Reply:
I've switched this to match on lower case, but im not sure how to handle the spaces and escape characters. We would need a utility to normalize table names ?.